core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+    return {
+      restrict: 'E',
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div><div></div>' +
+                '<div contenteditable="true" class="ckeditor-rte">' +
+                '</div></div>',
+      require: '?ngModel',
+      link: function(scope, el, attr, ngModel) {
+        var _RICH_TEXT_COMPONENTS = rteHelperService.getRichTextComponents();
+        var names = [];
+        var icons = [];
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!(scope.uiConfig() &&
+                scope.uiConfig().hide_complex_extensions &&
+                componentDefn.isComplex)) {
+            names.push(componentDefn.name);
+            icons.push(componentDefn.iconDataUrl);
+          }
+        });
+
+        // Rules to prevent CKEDITOR from filtering components and wrappers.
+        var componentFilterRules = names.map(function(name) {
+          return 'oppia-noninteractive-' + name;
+        }).join(' ') + '(*)[*];';
+        var wrapperFilterRules = ' div span[type];';
+
+        var pluginNames = names.map(function(name) {
+          return 'oppia' + name;
+        }).join(',');
+        var buttonNames = [];
+        names.forEach(function(name) {
+          buttonNames.push('Oppia' + name);
+          buttonNames.push('-');
+        });
+        buttonNames.pop();
+
+        // Initialize ckeditor.
+        var ck = CKEDITOR.inline(el[0].children[0].children[1], {
+          extraPlugins: 'widget,lineutils,sharedspace,' + pluginNames,
+          startupFocus: true,
+          extraAllowedContent: componentFilterRules + wrapperFilterRules,
+          sharedSpaces: {
+            top: el[0].children[0].children[0]
+          },
+          skin: 'bootstrapck,/third_party/static/ckeditor-bootstrapck/',
+          toolbar: [
+            {
+              name: 'history',
+              items: ['Undo', '-', 'Redo']
+            },
+            {
+              name: 'basicstyles',
+              items: ['Bold', '-', 'Italic', '-', 'RemoveFormat']
+            },
+            {
+              name: 'paragraph',
+              items: [
+                'NumberedList', '-',
+                'BulletedList', '-',
+                'Outdent', '-',
+                'Indent'
+              ]
+            },
+            {
+              name: 'rtecomponents',
+              items: buttonNames
+            }
+          ]
+        });
+
+        // A RegExp for matching rich text components.
+        var componentRe = (
+          /(<(oppia-noninteractive-(.+?))\b[^>]*>)[\s\S]*?<\/\2>/g
+        );
+
+        // Before data is loaded into CKEditor, we need to wrap every rte
+        // component in a span (inline) or div (block).
+        // For block elements, we add an overlay div as well.
+        var wrapComponents = function(html) {
+          if (html === undefined) {
+            return html;
+          }
+          return html.replace(componentRe, function(match, p1, p2, p3) {
```

Comment:
Good idea. I've decided to do this next to where I declare the regex, to avoid repetition.